1-Draft/RFCNNNN-Improve-Background-Operator.md

```diff
@@ -0,0 +1,216 @@
+with the command before it.
+
+### Fanning out to multiple background jobs and referencing the jobs with the `$!` variable
+
+```powershell
+$null = cmd1 & cmd2 & cmd3 &
+$!
+```
+
+```output
+Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
+--     ----            -------------   -----         -----------     --------             -------
+1      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man...
+2      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man...
+3      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man...
+```
+
+The `$!` variable in those two commands returns the jobs that were created with
+the command before it.
+
+### Launching a legacy command with the stop-parsing sigil and either the `&` or `&!` background operator
+
+This corrects a limitation with the background operator(s), which currently
+treat `&` (and `&!`) as an argument instead when used after the stop-parsing
+sigil.
+
+```powershell
+./plink.exe" --% $Hostname -l $Username -pw $Password $Command &
+```
+
+```output
+Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
+--     ----            -------------   -----         -----------     --------             -------
+1      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man...
+```
+
+### Showing or hiding the output from one or more jobs
+
+This allows job output to be shown or hidden on demand for any jobs.
+
+```powershell
+# Show the output from job ids 1-5
+Show-JobOutput -Id 1..5
+# Hide the output from job name Job2
+Hide-JobOutput -Name Job2
+# Show the output for the last job (or jobs) we launched
+Show-JobOutput -Job $!
+```
+
+```output
+# As described in the comments above, with the output showing in an overlay
+```
+
+### Viewing the process ID for a job
+
+This shows how you can access the read-only process ID for a job.
+
+```powershell
+$!.ProcessId
+```
+
+```output
+# The process id for the background job(s)
+```
+
+## Specification
+
+1. For jobs invoked with `&!` or `Start-Job -ShowOutput`, or for jobs referenced
+in `Show-JobOutput`, hook up event handlers on the output stream collections for
+the job objects to capture the output from the jobs as it is added in real
+time, and show that output in the current terminal with the job name or ID in
+front of each output record. Mirror how `ProgressInformation` is rendered when
+it comes to showing job output this way.
+1. Add a `&!` operator that starts a job with the output display turned on.
+1. Add a boolean property to `BackgroundJob` objects called `ShowOutput` that
+identifies if the job is currently configured to show output or not.
+1. Add an integer property to `BackgroundJob` objects called `ProcessId` that
```

Comment:
I'm not the expert here, but I'm having trouble figuring out if this needs to be added to `SMA.Job`, `SMA.PSRemotingJob` (the private type emitted by `Get-Job`), or `BackgroundJob` ([a static type we create in code](https://github.com/PowerShell/PowerShell/blob/bd6fdae73520931f0d27a29d6290e18761772141/src/System.Management.Automation/engine/remoting/commands/StartJob.cs#L26)). 

Also, would this just be blank if the Job object was not created as an "native executable" invocation but as a cmdlet invocation?

Reply:
I was looking at the [`S.M.A.Job.Output`, `.Error`, `.Warning`, `.Verbose`, `.Debug`, `.Information` and `.Progress` properties](https://github.com/PowerShell/PowerShell/blob/bd6fdae73520931f0d27a29d6290e18761772141/src/System.Management.Automation/engine/remoting/client/Job.cs#L1080-L1315), which are of type `PSDataCollection<PSStreamObject>`. Those collections have event handlers for `DataAdding` and `DataAdded`. Since those are on the abstract `S.M.A.Job` base class, I would expect this to be possible for any job.

Reply:
Whoops, I just re-read this, and realized you were specifically commenting about the `ProcessId` property. Right now I'm tending to lean to more/most specific rather than least specific for that, because while it would be useful for background jobs to surface a `ProcessId` property, it wouldn't be as useful for thread jobs, [where I just want the runspace ID exposed](https://github.com/PaulHigin/PSThreadJob/pull/41). With that in mind, either `BackgroundJob` or `S.M.A.PSRemotingJob` would probably be the right class to add that member to.

> Also, would this just be blank if the Job object was not created as an "native executable" invocation but as a cmdlet invocation?

If this member would only be available on jobs that were created as native executables, I don't think it would be present for other jobs. If we would prefer the property be available for any job, then we could either leave it blank for jobs running in the current process or show the current process id.

Reply:
I was unaware when I made this comment that jobs start as new processes by default. So PID would be definitely relevant. 

And some others have expressed in person that it'd be useful to have current PID for thread jobs as well.

Reply:
Updated RFC to have the `ProcessId` property for all jobs, including thread jobs, and to show it by default in output.